Python/chapter03/p05_sort_stack/miguelHx.py

```diff
@@ -0,0 +1,274 @@
+        n = self.top
+        while n.next is not None:
+            values.append(str(n.data))
+            n = n.next
+        values.append(str(n.data))
+        return '->'.join(values)
+
+
+class TestMyStack(unittest.TestCase):
+    def test_stack_push(self):
+        s = MyStack()
+        self.assertEqual(len(s), 0)
+        self.assertEqual(s.top, None)
+        s.push(2)
+        self.assertEqual(len(s), 1)
+        self.assertEqual(s.top.data, 2)
+        self.assertEqual(s.top.next, None)
+        s.push(3)
+        self.assertEqual(len(s), 2)
+        self.assertEqual(s.top.data, 3)
+        self.assertEqual(s.top.next.data, 2)
+        s.push(4)
+        self.assertEqual(len(s), 3)
+        self.assertEqual(s.top.data, 4)
+        self.assertEqual(s.top.next.data, 3)
+        self.assertEqual(list(s), [4, 3, 2])
+
+        # test adding different types, (float and int)
+        s.push(1.2)
+        self.assertEqual(len(s), 4)
+        self.assertEqual(s.top.data, 1.2)
+        self.assertEqual(s.top.next.data, 4)
+        self.assertEqual(list(s), [1.2, 4, 3, 2])
+
+    def test_stack_peek(self):
+        s = MyStack()
+        with self.assertRaises(IndexError):
+            s.peek()
+        s.push(1)
+        s.push(2)
+        s.push(99)
+        top_val = s.peek()
+        self.assertEqual(top_val, 99)
+
+    def test_stack_pop(self):
+        # first case, attempt to pop an empty stack
+        s = MyStack()
+        with self.assertRaises(IndexError):
+            s.pop()
+        s.push(1)
+        s.push(2)
+        s.push(3)
+        # size is 3
+        self.assertEqual(list(s), [3, 2, 1])
+        val = s.pop()
+        self.assertEqual(val, 3)
+        self.assertEqual(s._size, 2) # size should now be 2
+        self.assertEqual(list(s), [2, 1])
+
+    def test__bool__(self):
+        s = MyStack()
+        self.assertFalse(s)
+        s.push(3)
+        self.assertTrue(s)
+
+
+def sorted_stack(stack: MyStack[T]) -> MyStack[T]:
+    """This function will take in a stack
+    and return a sorted copy.
+    In order to do this, we need to find the biggest
+    number in an input stack copy, and then push it to the
+    output stack. Repeat until input stack copy is empty.
+
+    Args:
+        stack (MyStack): stack of items
+
+    Returns:
+        MyStack: sorted stack of items, with smallest items on top
+    """
+    # create empty output stack
+    output_stack: MyStack = MyStack()
+    # make copy of input stack.
+    stack_copy: MyStack = stack.produce_copy()
+    # create temporary auxiliary stack
+    aux_stack: MyStack = MyStack()
+    # we will extract max values until stack_copy is empty.
+    while len(stack_copy) > 0:
+        # look for index of max item in stack_copy
+        max_index = 0
+        max_value: T = stack_copy.peek()
+        item: T
+        for i, item in enumerate(stack_copy):
+            if item > max_value:
+                max_value = item
+                max_index = i
```

Comment:
note that this is also forbidden by the rules of the game: stacks can only be interacted with through `push()`, `peek()`, and `pop()` and checking for emptiness (with `while stack_copy:`)

Reply:
Re-implemented here, this time following the rules of the game:  ebee21966031a34637f2b47c0eaabf9d96ca8d5d